examples/gno.land/p/demo/accesscontrol/accesscontrol.gno

```diff
@@ -0,0 +1,156 @@
+// import "gno.land/p/demo/accesscontrol"
+//
+// Create a new role with a specific admin.
+// adminRole := std.Address("admin-address")
+// role := accesscontrol.NewRole("ExampleRole", adminRole)
+//
+// Check if an account has a specific role.
+// account := std.Address("user-address")
+// hasRole := role.HasRole(account)
+//
+// Grant a role to a specific account.
+// role.GrantRole(account)
+//
+// Revoke a role from a specific account.
+// role.RevokeRole(account)
+//
+// Renounce a role with caller confirmation.
+// role.RenounceRole(std.GetOrigCaller())
+//
+// Change the admin role for a specific role.
+// newAdmin := std.Address("new-admin-address")
+// role.SetRoleAdmin(newAdmin)
+package accesscontrol
+
+import (
+	"errors"
+	"std"
+
+	"gno.land/p/demo/avl"
+)
+
+var ErrUnauthorized = errors.New("unauthorized; caller is not admin")
+
+// Role struct to store role information
+type Role struct {
+	Name      string
+	Holders   *avl.Tree // -> std.Address -> bool
+	AdminRole std.Address
+}
+
+type Roles struct {
+	AllRoles []*Role
+	Admin    std.Address
+}
+
+// NewRole creates a new instance of Role
+func NewRole(name string, adminRole std.Address) *Role {
+	return &Role{
+		Name:      name,
+		Holders:   avl.NewTree(),
+		AdminRole: adminRole,
+	}
+}
+
+// Method to check if the caller has the admin role and return an error
+func (r *Role) CallerIsAdmin() error {
+	caller := std.PrevRealm().Addr()
+
+	if r.AdminRole != caller {
+		return ErrUnauthorized
+	}
+
+	return nil
+}
+
+// Method to assert if the caller has the admin role, panics if not
+func (r *Role) AssertCallerIsAdmin() {
+	if err := r.CallerIsAdmin(); err != nil {
+		panic(err)
+	}
+}
+
+// Method to create a new role within the realm
+func (rs *Roles) CreateRole(name string, adminRole std.Address) *Role {
```

Comment:
why is the second parameter named `adminRole`?

Reply:
```go
if rs.Admin != std.PrevRealm().Addr() {
		panic("accesscontrol: caller does not have the global admin role")
	}
```
should `panic` be replaced by `ErrUnauthorized`, and when is the use of panic in a package necessary?

Reply:
> ```go
> if rs.Admin != std.PrevRealm().Addr() {
> 		panic("accesscontrol: caller does not have the global admin role")
> 	}
> ```
> 
> should `panic` be replaced by `ErrUnauthorized`, and when is the use of panic in a package necessary?

I think it's better to use `panic` because to call this `function`, we require the caller to be the global administrator, and if they are not, it's a critical error